mssqlvdi: set default for serveraddress+instance on restore

arogge opened the pull request:
Previously, mssqlvdi set a default for serveraddress  & instance on backup, but not
on restore. If you did not specify serveraddress on backup, restore
would fail unless you specified it in the plugin options manually.

This patch now sets the default for serveraddress&instance on restore just like
it does on backup.

### Thank you for contributing to the Bareos Project!

#### Please check

- [x] Short description and the purpose of this PR is present _above this paragraph_
- [x] Your name is present in the AUTHORS file (optional)

If you have any questions or problems, please give a comment in the PR.

### Helpful documentation and best practices

- [Git Workflow](https://docs.bareos.org/DeveloperGuide/gitworkflow.html)
- [Automatic Sourcecode Formatting](https://docs.bareos.org/DeveloperGuide/generaldevel.html#automatic-sourcecode-formatting)
- [Check your commit messages](https://docs.bareos.org/DeveloperGuide/gitworkflow.html#commits)


### Checklist for the _reviewer_ of the PR (will be processed by the Bareos team)
Make sure you check/merge the PR using `devtools/pr-tool` to have some simple automated checks run and a proper changelog record added.

##### General
- [x] Is the PR title usable as CHANGELOG entry?
- [x] Purpose of the PR is understood
- [x] Commit descriptions are understandable and well formatted
- [x] Check backport line
- [x] Required Backport has been created

##### Source code quality
- [x] Source code changes are understandable
- [x] Variable and function names are meaningful
- [x] Code comments are correct (logically and spelling)
- [x] Required documentation changes are present and part of the PR

##### Tests
- [x] Decision taken that a test is required (if not, then remove this paragraph)
- [x] The choice of the type of test (unit test or systemtest) is reasonable
- [x] Testname matches exactly what is being tested
- [x] On a fail, output of the test leads quickly to the origin of the fault

arogge committed:
mssqlvdi: set default for serveraddress on restore

Previously, mssqlvdi set a default for serveraddress on backup, but not
on restore. If you did not specify serveraddress on backup, restore
would fail unless you specified it in the plugin options manually.

This patch now sets the default for serveraddress on restore just like
it does on backup.

sebsura commented on core/src/win32/plugins/filed/mssqlvdi-fd.cc:
```diff
@@ -958,6 +958,14 @@ static void SetAdoConnectString(PluginContext* ctx)
   PoolMem ado_connect_string(PM_NAME);
   plugin_ctx* p_ctx = (plugin_ctx*)ctx->plugin_private_context;
 
+  // If no explicit instance name given usedthe DEFAULT_INSTANCE name.
```
Typo
```suggestion
  // If no explicit instance name given use the DEFAULT_INSTANCE name.
```

sebsura commented:
I updated the pr template.

sebsura commented on core/src/win32/plugins/filed/mssqlvdi-fd.cc:
```diff
@@ -958,6 +958,14 @@ static void SetAdoConnectString(PluginContext* ctx)
   plugin_ctx* p_ctx = (plugin_ctx*)ctx->plugin_private_context;
 
+  // If no explicit instance name given usedthe DEFAULT_INSTANCE name.
+  if (!p_ctx->instance) { p_ctx->instance = strdup(DEFAULT_INSTANCE); }
+
+  // If no explicit server address given use the DEFAULT_SERVER_ADDRESS.
+  if (!p_ctx->server_address) {
+    p_ctx->server_address = strdup(DEFAULT_SERVER_ADDRESS);
+  }
```
I'm not sure about this, but i think this is done somewhere else.  I think its kinda unexpected that a function called `SetAdoConnectString` actually also sets the server_address and the instance.

I had the two following ideas:
1. since `parse_plugin_definition` kinda acts like a constructor we could add 
``` 
  SetStringIfNull(&p_ctx->server_address, DEFAULT_SERVER_ADDRESS);
  SetStringIfNull(&p_ctx->instance, DEFAULT_INSTANCE);
``` 
  just before returning.  This has the obvious downside that it can change the behaviour of that function if it ever gets called twice on the same context.   I hope this does not actually happen!

2. Add `get_server_name` and `get_instance` functions that return the default if the real ones are set to null.  This has the advantage that we do not need to needlessly copy the default values.  In that case the functions should return a `const char *` just to be safe.  This comes with the downside of breaking if the rest of the code expects `p_ctx->instance` to not be zero.

What do you think about these suggestions ?

sebsura replied:
Looking at the code (in this file at least) it seems that it would be pretty easy to replace accesses with calls to the accessor function instead. So maybe it could work.

sebsura commented:
I discovered a new issue that is maybe related to this:

When i try to restore a backup (full + incr) with this plugin and specify pluginoptions (but not `instance`!), then the fd crashes.  I.e.
```
restore client=bareos-fd fileset=DBNormalTest where=/ pluginoptions=mssqlvdi:serveraddress=<address>:database=DBNormalTest:norecovery=yes:replace=yes:recoverafterrestore=yes select all done yes
```
leads to a crash

sebsura commented:
So Im assuming we also do not set `instance` to the default value on restore

sebsura commented:
Looking into your fixes, this is actually already included here.  The pr simply does not mention it :)

bruno-at-bareos commented:
We want this merge just after #2072

sebsura committed:
tests: add mssql test that tests default setting

sebsura committed:
mssqlvdi: fix nullptr accesses

During restore SetupVdiDevice was still getting called with
instance=nullptr.  This caused the fd to crash when it tries to memcpy
it.

To prevent this completely, all accesses to this member (and server
address) are now done through a function that does the default setting
by itself.

sebsura commented on systemtests/tests/mssqlvdi-plugin/testrunner-06-DBNormal-DropDB:
```diff
@@ -0,0 +1,45 @@
+if "${SQLCMD}" -Q "SELECT * FROM DBNormalTest.tests.samples" | grep Bareos; then
+  echo "Database is not empty"
+  exit 2
+fi
```
Could you add a comment explaining what this command does to the proposed fix ?

sebsura replied:
fixed.

sebsura committed:
mssql: make the test for dropped db more accurate

sebsura commented on systemtests/tests/mssqlvdi-plugin/testrunner-16-DBFileStream-DropDB:
```diff
@@ -0,0 +1,45 @@
+if "${SQLCMD}" -Q "SELECT * FROM DBFileStreamTest.tests.samples" | grep Bareos; then
+  echo "Database is not empty"
+  exit 2
+fi
```
fixed.

pstorz committed:
mssqlvdi test: harden mssql drop verification

pstorz reviewed (approved):
Looks good

sebsura committed:
docs: adjust the mssql docs to the new behaviour

sebsura commented on docs/manuals/source/Appendix/Howtos/BackupOfThirdPartyDatabases.rst.inc:
```diff
@@ -451,16 +467,17 @@ Example for a full restore:
 
 recoverafterrestore=<yes|no>
-   With this command the database is right after backup in the correct mode. If you not use this you have to use the followed tsql statement:
+   Automatically issues a recover after bareos is done with all its restores. If you not use this you have to use the followed tsql statement:
```
rewritten the text

sebsura commented on docs/manuals/source/Appendix/Howtos/BackupOfThirdPartyDatabases.rst.inc:
```diff
@@ -305,14 +296,39 @@ If the *where* parameter is set **to a directory** instead of '/', the plugin wi
+   old name.  This means that if these files still exist and are used, then the restore under the new name
+   will fail, even if you set ``replace=yes``.
+   If the files exist because an old version of that database still exists, then you can restore the database
+   by restore to the old name with ``replace=yes`.  Otherwise you can restore the db via the local files approach.
```
fixed

pstorz reviewed (approved):
Looks good!

BareosBot committed:
Update CHANGELOG.md
